src/trajectory.jl

```diff
@@ -246,7 +286,7 @@ function randcat(rng, zs::AbstractVector{<:PhasePoint}, unnorm_ℓP::AbstractMat
 end
 
 function samplecand(rng, τ::StaticTrajectory{MultinomialTS}, h, z)
-    zs = step(τ.integrator, h, z, τ.n_steps; res=[z for _ in 1:abs(τ.n_steps)])
+    zs = step(τ.integrator, h, z, τ.n_steps; full_trajectory = Val(true))
```

Comment:
> Maybe just make it step(...; full_trajectory::Bool = false) instead? Julia's constant propagation should compile it away in the cases that we use it anyways since it's always hard-coded into the method depending on what it needs.

If this true I'd prefer `full_trajectory=true`.

Reply:
Hmm, yeah it's a bit unclear to me whether or not it works.

For example, if I call `samplecand` I get the following:
```julia
julia> @code_warntype AHMC.samplecand(rng, proposal, hamiltonian, z₀)
Variables
  #self#::Core.Compiler.Const(AdvancedHMC.samplecand, false)
  rng::Core.Compiler.Const(Random._GLOBAL_RNG(), false)
  τ::StaticTrajectory{MultinomialTS,Leapfrog{Float64}}
  h::Hamiltonian{DiagEuclideanMetric{Float64,Array{Float64,1}},typeof(ℓπ),AdvancedHMC.var"#∂ℓπ∂θ#43"{typeof(ℓπ)}}
  z::AdvancedHMC.PhasePoint{Array{Float64,1},AdvancedHMC.DualValue{Float64,Array{Float64,1}}}
  zs::Union{AdvancedHMC.PhasePoint{Array{Float64,1},AdvancedHMC.DualValue{Float64,Array{Float64,1}}}, Array{AdvancedHMC.PhasePoint{Array{Float64,1},AdvancedHMC.DualValue{Float64,Array{Float64,1}}},1}}
  ℓws::Array{Float64,1}
  unnorm_ℓprob::Array{Float64,1}

Body::AdvancedHMC.PhasePoint{Array{Float64,1},AdvancedHMC.DualValue{Float64,Array{Float64,1}}}
1 ─       Core.NewvarNode(:(unnorm_ℓprob))
│   %2  = Base.getproperty(τ, :integrator)::Leapfrog{Float64}
│   %3  = Base.getproperty(τ, :n_steps)::Int64
│         (zs = AdvancedHMC.step(%2, h, z, %3, true))
│   %5  = Base.broadcasted(AdvancedHMC.energy, zs)::Base.Broadcast.Broadcasted{Base.Broadcast.DefaultArrayStyle{1},Nothing,typeof(AdvancedHMC.energy),Tuple{Array{AdvancedHMC.PhasePoint{Array{Float64,1},AdvancedHMC.DualValue{Float64,Array{Float64,1}}},1}}}
│   %6  = Base.materialize(%5)::Array{Float64,1}
│         (ℓws = -%6)
│   %8  = AdvancedHMC.eltype(ℓws)::Core.Compiler.Const(Float64, false)
│   %9  = (%8 <: AdvancedHMC.AbstractVector)::Core.Compiler.Const(false, false)
└──       goto #3 if not %9
2 ─       Core.Compiler.Const(:(ℓws = Core._apply_iterate(Base.iterate, AdvancedHMC.hcat, ℓws)), false)
3 ┄       (unnorm_ℓprob = ℓws)
│   %13 = AdvancedHMC.randcat(rng, zs, unnorm_ℓprob)::AdvancedHMC.PhasePoint{Array{Float64,1},AdvancedHMC.DualValue{Float64,Array{Float64,1}}}
└──       return %13
```
So there's a `Union` inferred for `zs` which means that the type is not inferred, i.e. constant-propagation isn't working, despite the fact that we can see that `AdvancedHMC.step` is called with `true` as the last argument (I just made it an argument rather than kwarg because I think kwargs aren't "compiled" away as aggressively).

It also doesn't seem to work if I wrap the `step` call:
```julia
julia> a(integrator, hamiltonian, z₀) = AHMC.step(integrator, hamiltonian, z₀, 2, true)
a (generic function with 2 methods)

julia> @code_warntype a(integrator, hamiltonian, z₀)
Variables
  #self#::Core.Compiler.Const(a, false)
  integrator::Leapfrog{Float64}
  hamiltonian::Hamiltonian{DiagEuclideanMetric{Float64,Array{Float64,1}},typeof(ℓπ),AdvancedHMC.var"#∂ℓπ∂θ#43"{typeof(ℓπ)}}
  z₀::AdvancedHMC.PhasePoint{Array{Float64,1},AdvancedHMC.DualValue{Float64,Array{Float64,1}}}

Body::Union{AdvancedHMC.PhasePoint{Array{Float64,1},AdvancedHMC.DualValue{Float64,Array{Float64,1}}}, Array{AdvancedHMC.PhasePoint{Array{Float64,1},AdvancedHMC.DualValue{Float64,Array{Float64,1}}},1}}
1 ─ %1 = AdvancedHMC.step::Core.Compiler.Const(AdvancedHMC.step, false)
│   %2 = (%1)(integrator, hamiltonian, z₀, 2, true)::Union{AdvancedHMC.PhasePoint{Array{Float64,1},AdvancedHMC.DualValue{Float64,Array{Float64,1}}}, Array{AdvancedHMC.PhasePoint{Array{Float64,1},AdvancedHMC.DualValue{Float64,Array{Float64,1}}},1}}
└──      return %2
```

However, it _does_ work if I wrap `step` using `const` variables:
```julia
julia> a() = AHMC.step(integrator_const, hamiltonian_const, z_const, 2, true)
a (generic function with 1 method)

julia> @code_warntype a()
Variables
  #self#::Core.Compiler.Const(a, false)

Body::Array{AdvancedHMC.PhasePoint{Array{Float64,1},AdvancedHMC.DualValue{Float64,Array{Float64,1}}},1}
1 ─ %1 = AdvancedHMC.step::Core.Compiler.Const(AdvancedHMC.step, false)
│   %2 = (%1)(Main.integrator_const, Main.hamiltonian_const, Main.z_const, 2, true)::Array{AdvancedHMC.PhasePoint{Array{Float64,1},AdvancedHMC.DualValue{Float64,Array{Float64,1}}},1}
└──      return %2
```

Therefore I hereby summon the two gods of Julia optimization, oh @mohamed82008 and @devmotion , please provide us peasants with thoust holy guidance!

Reply:
Alternatively, we just leave it as `Val` for now:)

Reply:
Maybe you observe https://discourse.julialang.org/t/how-does-constant-propagation-work/22735/4?u=devmotion here.

Reply:
Haha, I just edited the comment to add that link too:) 

Doesn't make much wiser as to what I should do though. Do I leave as is or do you swap to `Bool` and cross my fingers + make it so that I can't inspect the type-stability of calls involving `step`? :confused: 

Feel like leaving it with `Val` is okay for now

Reply:
Looks good to me - we can improve this in future PR if necessary.

Reply:
I don't think constant propagation works with kwargs, not sure though.